### DISCUSSION TITLE
一个插件 idea

### DISCUSSION THREAD

**Discussion by H4M5TER** [koishijs/koishi#236] [Ideas]
我之前有给 bot 做b站直播推送和动态推送的功能，然后不知道怎么拆出来。今天阅读了一下 koishi-plugin-rss 的代码，感觉可以把 rss 的部分拆开，变成 koishi-plugin-subscribe (随便乱取的名字) ，可能只负责注册指令调用子插件给出（怎么做?）的订阅取订接口，然后独立出 koishi-plugin-subscribe-rss 作为子插件。

**Comment by H4M5TER** [comment]
@Shigma

**Comment by shigma** [comment]
聚合推送我之前考虑过。我一开始想的是像 github，gitlab，rss 都可以整合起来做。但是后来没有做，因为 github 这方面做了非常细致的特化。其实现在回头看的话，一旦写这种东西，有两个显著的坏处：

1. 你没法对各个协议做特殊的支持，只能被动地设计一个通用接口，这样支持必然就会弱，大家还是会各用各的（类比 relay 功能，common 插件中有个 relay 配置项，但是大家还是会自己写 relay 逻辑，并不能起到作用）
2. 每个协议单独实现，会让包的数量显著增加，这中间的中转逻辑对 Koishi 来说完全不值得（我目前的计划是 Koishi 官仓只维护复杂的核心插件，所以不会搞一个这种东西）

但上面说的是对我自己而言。你依然可以这么做。不过在做之前你或许确实应该考虑一下 RSS Hub。有信心做的比它更好吗？如果没有的话，为什么不直接去优化 RSS 插件去适应 RSS Hub 的输出呢？

**Reply by H4M5TER** [reply]
1. 应该说比较难对每个协议做独立的支持，但仍然是有可行性的
2.
- 没问题，是打算单独维护一个
- 中转逻辑指的是什么？
3. 我了解过 RSS Hub，它是一个由社区维护的，一系列拉取不定格式的源转化为 RSS 订阅源的工具集合，不太明白有信心比它更好是指什么，聚合和推送的操作其实还是在 RSS 插件里做的，如果使用 RSS Hub + RSS 插件来聚合推送，会存在很多问题：
- 我首先需要为 RSS Hub 编写订阅源，对于一些符合 RSS 性质（例如博客新闻什么的，这点或许是我刻板印象）但没有提供 RSS 订阅的源，我的确乐意为 RSS Hub 编写订阅源，但从编写到合并再到能够使用有（较为严重的）时间上的延迟。
- 从源到 RSS Hub 再到 RSS 插件，存在较多的网络传输，增加了故障的可能性（虽然，应该确实没多少），经过了更多层，增加了处理故障（不光是网络故障，也包括 RSS Hub 和 RSS 插件两端的程序故障）的困难。
- 正如你说的，“你没法对各个协议做特殊的支持，只能被动地设计一个通用接口”，RSS 插件也是如此，如果要做一些特殊的支持，例如对信息做一些客制化的处理，要么在 RSS 插件里做一些针对性的判断，要么在 RSS Hub 的订阅源处做一些处理。前者会使 RSS 插件变得不纯粹，后者会增加用户的使用门槛。
4. 我想做的到底是什么，我希望有一个 subscribe 插件，让我可以 `订阅 <类型> <配置>` 这样订阅各种东西，具体有什么类型的东西由安装的子插件决定。同时希望（对于每个子插件）能有较高的可配置性，例如实现一个 steam 拉取愿望单的子插件，我可能希望愿望单里的游戏打折了提醒，或是愿望单里尚未推出的游戏推出时提醒，或是有史低/新史低价时提醒。
5. 但具体 subscribe 插件和子插件之间怎么交互（例如，子插件应该暴露 subscribe 和 unsubscribe 接口，以及自身的一些信息给主插件，应如何做？），我还没有一个好的思路，如果能给一些帮助或是拍醒一些不实际的地方，我会非常感激。

**Comment by shigma** [comment]
有关插件如何暴露全局的接口，有两种思路：

第一种可以参考 puppeteer 插件。直接在 app 上面赋值即可：

```ts
declare module 'koishi-core' {
  interface App {
    browser?: Browser
  }
}

export function apply(ctx: Context) {
  // 只能通过 app 访问
  ctx.app.browser = await launch(options)
}
```

第二种是使用 `Context.delegate()` 方法直接暴露在所有上下文的原型链中。比如 webui 就采用了这种方法：

```ts
declare module 'koishi-core' {
  interface Context {
    webui?: WebUIServer
  }
}

Context.delegate('webui')

export function apply(ctx: Context) {
  // 可以在任何 ctx 上访问
  ctx.webui = new WebUIServer()
}
```

包括大家熟悉的 `ctx.router`, `ctx.database` 也是使用了这个函数。

显然是后一种方法看起来比较好，不过这种写法也对这里的实例有一定的要求。不理解 delegate 原理的话，非常容易写挂（很多时候是因为 this 指代不清）。如果遇到问题可以在这里问我。

**Reply by H4M5TER** [reply]
指南里面可以补充“插件如何暴露全局的接口”到插件相关章节，API文档里没有提到 Context.delegate()，也需要补充。
希望能陈述一下 delegate 原理以及对实例的要求。（实例指什么？Context？）

**Reply by shigma** [reply]
这个我其实是因为不知道该补到哪一页……
